genie-agent/src/main/java/com/netflix/genie/agent/execution/services/DownloadServiceImpl.java

```diff
@@ -0,0 +1,195 @@
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+
+package com.netflix.genie.agent.execution.services;
+
+import com.google.common.collect.Maps;
+import com.google.common.collect.Sets;
+import com.netflix.genie.agent.execution.exceptions.DownloadException;
+import lombok.Getter;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.springframework.context.annotation.Lazy;
+import org.springframework.stereotype.Component;
+
+import javax.annotation.Nullable;
+import java.io.File;
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * Cache-backed download service to retrieve a set of files and place them at an expected location, according
+ * to a given manifest.
+ *
+ * @author mprimi
+ * @since 4.0.0
+ */
+@Component
+@Lazy
+@Slf4j
+class DownloadServiceImpl implements DownloadService {
+    private final FetchingCacheService fetchingCacheService;
+
+    DownloadServiceImpl(final FetchingCacheService fetchingCacheService) {
+        this.fetchingCacheService = fetchingCacheService;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public Manifest.Builder newManifestBuilder() {
+        return new ManifestImpl.Builder();
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void download(final Manifest downloadsManifest) throws DownloadException {
+
+        // Validate all destination directories exist
+        for (final File targetDirectory : downloadsManifest.getTargetDirectories()) {
+            if (!targetDirectory.exists()) {
+                throw new DownloadException(
+                    "Target directory does not exist: " + targetDirectory.getAbsolutePath()
+                );
+            } else if (!targetDirectory.isDirectory()) {
+                throw new DownloadException(
+                    "Target directory is not a directory: " + targetDirectory.getAbsolutePath()
+                );
+            }
+        }
+
+        // Validate all target files don't exist
+        for (final File targetFile : downloadsManifest.getTargetFiles()) {
+            if (targetFile.exists()) {
+                throw new DownloadException("Target file exists: " + targetFile.getAbsolutePath());
+            }
+        }
+
+        try {
+            fetchingCacheService.get(downloadsManifest.getEntries());
+        } catch (final IOException e) {
+            throw new DownloadException("Failed to download", e);
+        }
+    }
+
+    @Getter
```

Comment:
Do you want `EqualsAndHasCode` for storing in collections and `ToString` for debugging?

Reply:
I don't expect to have collections of manifests, tho wouldn't hurt to add those.
`ToString` would be nice to print the map as JSON...